### DISCUSSION TITLE
Модернизация системы `TargetFramework`

### DISCUSSION THREAD

**Discussion by samuraiGH** [pascalabcnet/pascalabcnet#3068] [General]
### Проблематика
Пока буду рассматривать проблему в контексте `.NET Framework`
На данный момент компилятор все необходимые для компиляции сборки подгружает из `%SYSTEMDRIVE%\Windows\Microsoft.NET\assembly\GAC_MSIL`. Сборки, расположенные по данному пути, соответствуют самой высокой установленной версии `.NET Framework`.

Таким образом, компилятор никак не может помочь нам, если у нас установлен `net48`, а мы хотим чтобы наша программа гарантированно запускалась например в `net46`. Мы легко можем использовать классы, которые есть в `net48`, но отсутствуют в `net46`. Единственный вариант избежать этого -- проверять все используемые API по документации. Однако это неправильно, компилятор должен сам контролировать такие вещи.

### Что с этим можно сделать?
Ключевой момент -- использование вместо Implementation Assemblies (которые используются сейчас), так называемых [Reference Assemblies](https://learn.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies). Они имеются для каждой версии `.NET Framewrk` и устанавливаются вместе с Developer Pack в папку `%ProgramFiles(x86)%\Reference Assemblies\Microsoft\Framework\.NETFramework`
Reference Assemblies не перекрывают друг друга, в отличие от Implementation Assemblies и мы всегда можем чётко сказать для какой версии `.NET Framework` мы собираем программу

### Как это сделать?
1. Если директива `TargetFramework` не прописана, то компилятор может работать так, как он работает сейчас;
2. Если `TargetFramework` прописан внутри `program`, то компилятор подгружает библиотеки из соответствующей папки. В том числе `mscorlib.dll` и все, что прописаны в `{$reference}`, если они относятся к стандартным;
3. Если `TargetFramework` прописан внутри `unit`, то компилятор работает аналогично с пункту №2. Модуль не должен скомпилироваться, если используется недоступные в этой версии API. Это в том числе позволит защитить PABCSystem и прочие стандартные модули от небезопасных изменений, нарушающих совместимость с `net40`;
4. Что если в `program` выставлен `net48`, а в подключенном модуле `net46`? В модуле не должны использоваться недоступные в `net46` API, а вот в `program` они вполне могут использоваться. В результирующем `exe` `TargetFramewrkAttribute` должен быть `net48`;
5. Что если наоборот в `program` выставлен `net46`, а в подключенном модуле `net48`? Модуль можно скомпилировать в pcu, а основная программа компилироваться не должна;
6. Компилятор должен разворачивать кеш-функции, учитывая `TargetFramework`. Если `TargetFramework` ниже `net47`, то нужно использовать `Tuple` вместо `ValueTuple`

### Некоторые детали
Не под каждую версию `.NET Framework` есть Developer Pack. Для вышедших из поддержки версий можно скачать Reference Assemblies отдельно из NuGet. На описанное выше это не влияет, но знать полезно

При работе под Linux путь для `Reference Assemblies` разумеется другой. Нужно будет найти стандартный

**Comment by miks1965** [comment]
Еще бы найти тех, кто будет всё это реализовывать для устаревшей NET Framework

**Comment by SunSerega** [comment]
Это сейчас только для .Net Framework, но эту систему будет легко расширить до .Net Core, когда компилятор будет с ним совместим.

**Comment by miks1965** [comment]
Ну ОК, ждём реализаторов

**Comment by ibond84** [comment]
В .NET Core принято таскать рантайм с собой или, если возможно, компилировать все в один нативный exe-шник

**Comment by SunSerega** [comment]
"принято" существует в определённых контекстах.

Одно из основных преимуществ .Net это портабельность и маленький размер программ, потому что многое уже прописано в среде выполнения.
Есть большие проекты в которых размер рантайма по сравнению со всем остальным это мелочь.
Есть проекты которые ожидаемо разворачиваются на своей индивидуальной виртуальной машине (к примеру бэкэнд сайта), где только они сами будут использовать среду .Net - можно сразу и у версию указать явно.
Есть игры на движках вроде юнити, где в принципе "принято" брать самое простое но тупое и неэффективное решение.

Эти все проекты большие и громкие, но с текущим набором инструментов - область применения PascalABC.Net с ними практически не пересекается.
А в мелких программах, вроде [CKAN](https://github.com/KSP-CKAN/CKAN), или [моего последнего проекта](https://github.com/SunSerega/temp-Mandelbrot-example), важнее то что можно взять .exe-шник со страницы релиза и просто запустить его. На любой OS, в любой папке (хоть на рабочем столе) и т.п.

**Comment by samuraiGH** [comment]
> В .NET Core принято таскать рантайм с собой

Нет. `SelfContained` это только фишка, которая к тому же имеет свои проблемы. Отсутствие кроссплатформенности это первая из них.
Для бэкенда допустим в ней смысла вообще нет. 
Да и для десктопа может быть удобнее установить рантайм вместе с основной программой. Мне например не составляет проблем пару лишних действий в установщике прописать. Если установщик и так есть, то какой смысл запихивать рантайм в exe, да ещё и плодить несколько exe под каждую платформу?

Да и проблему это никак не решает. Версия может быть ограничена по принципу LTS. К тому же в дотнете тоже есть breaking changes и если при обновлении что-то отвалится из-за неявного перехода на новую версию -- будет совсем глупо. Да и в библиотеку рантайм не притянешь

**Comment by miks1965** [comment]
Да, правильно - спасибо что конвертировали в дискуссию

**Reply by samuraiGH** [reply]
Да нет. Это вполне конкретная проблема. И лучше бы она была на виду, чтобы внешние разработчики были в курсе и могли включиться по возможности

**Comment by ghost** [comment]
This reference assemblies vs. implementation assemblies caused many troubles for me:

https://github.com/bflattened/bflat/issues/171

I hate .NET Core. I prefer .NET Framework, the .NET that I used to know.
